### PR TITLE
Read cluster domain from resolv.conf or env

### PR DESCRIPTION
**Pull request by harryge00**
## Why are these changes needed?

1. In k8s,  [cluster domains](https://kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/#introduction) are configuration.  But now kuberay hard coded it to `cluster.local`. So fix this to either read cluster domain from `/etc/resolv.conf` or environmental variable

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests

### CODE REVIEW

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (Jeffwan) [inline ray-operator/controllers/ray/utils/dashboard_httpclient.go:124]: This is reasonable!

**Review by Jeffwan — APPROVED** [review]

**Comment by kevin85421** [comment]
Thanks @harryge00 for pointing this out! If possible, please provide a detailed reproduction script for either Kind or Minikube so that I can also test it manually. Your assistance in this matter would be greatly appreciated.

**Comment by harryge00** [comment]
> Thanks @harryge00 for pointing this out! If possible, please provide a detailed reproduction script for either Kind or Minikube so that I can also test it manually. Your assistance in this matter would be greatly appreciated.↳


@kevin85421 

```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  kubeadmConfigPatches:
  - |
    kind: ClusterConfiguration
    metadata:
      name: config
    networking: 
      dnsDomain: "foo.bar.com"
  - |
    kind: InitConfiguration
    nodeRegistration:
      kubeletExtraArgs:
        cluster-domain: "foo.bar.com"
```
`kind create cluster --config=config.yaml` then a k8s cluster will be created with `foo.bar.com` domain name.

Deploy kuberay:
https://github.com/ray-project/kuberay#use-yaml

Deploy rayjob:
```
kubectl apply -f ray-operator/config/samples/ray_v1alpha1_rayjob.yaml
```

Get rayjob status:
```
kubectl get rayjob -oyaml
```
image
Check the dashboardURL.

File: ray-operator/controllers/ray/utils/util.go
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util.go:35]: nit
```suggestion
	ClusterDomainEnvKey = "CLUSTER_DOMAIN"
```

File: ray-operator/controllers/ray/utils/util.go
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util.go:36]: nit
```suggestion
	DefaultDomainName   = "cluster.local"
```

File: ray-operator/controllers/ray/utils/util.go
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util.go:175]: Good catch!

**Review by kevin85421 — COMMENTED** [review]
Thanks @harryge00 for this contribution! I recently met a lot of networking issues reporting by users (https://github.com/ray-project/kuberay/issues/955). This PR can prevent some users from networking issues in the future!

One of my concern about this PR is the complexity. This PR is a bit complex and risky for maintainers because it directly parses `/etc/resolv.conf`. 

   * Maintainers need to have related context about DNS and the specification of `resolv.conf`.
   * The operator does too many things for users. It may introduces some "unknown unknowns" in some edge cases.

My recommendation is to keep it simple. That is, making the value of domain name configurable (i.e. `CLUSTER_DOMAIN`) and documenting the usage are good enough.

By the way, I have sent you a direct message in Slack. We can have a 15-mins chat on Zoom. We can accelerate the review process and make this PR be merged earlier. Again, thanks for your contribution!

**Comment by harryge00** [comment]
@kevin85421 I have updated the code, to use env only. Where should I add the document to? I think there should be a document recording all the parameters/env vars for the operator?

**Comment by kevin85421** [comment]
> @kevin85421 I have updated the code, to use env only. Where should I add the document to? I think there should be a document recording all the parameters/env vars for the operator?

Thank you for the update! You can create a document in this [directory](https://github.com/ray-project/kuberay/tree/master/docs/guidance). We have an issue to track the progress of the document for environment variables (https://github.com/ray-project/kuberay/issues/957), and we will finish it before release v0.5.0.

**Review by kevin85421 — APPROVED** [review]
LGTM. I will merge it after I test it manually tomorrow. Thank you for this contribution!

**Comment by kevin85421** [comment]
I test this PR manually.

* Step 1: Create a Kind cluster by following this https://github.com/ray-project/kuberay/pull/951#issuecomment-1465098772
* Step 2: Add a environment variable `CLUSTER_DOMAIN` in KubeRay operator
  ```yaml
  env:
  - name: CLUSTER_DOMAIN
    value: foo.bar.com
  ```

* Step 3: Deploy a RayJob
  * Note that: we need to update the init container. Currently. the init container hardcoded the CLUSTER_DOMAIN with `cluster.local`.
    ```yaml
    initContainers:
      # the env var $FQ_RAY_IP is set by the operator if missing, with the value of the head service name
      - name: init
        image: busybox:1.28
        command: [ 'sh', '-c', "until nslookup $RAY_IP.$(cat /var/run/secrets/kubernetes.io/serviceaccount/namespace).svc.cluster.local; do echo waiting for K8s Service $RAY_IP; sleep 2; done" ]
    # Replace: "until nslookup $RAY_IP.$(cat /var/run/secrets/kubernetes.io/serviceaccount/namespace).svc.cluster.local;"
    # With: "until nslookup $FQ_RAY_IP"
    ```
  ```
  kubectl apply -f ray_v1alpha1_rayjob.yaml
  ```
  Screen Shot 2023-03-20 at 3 35 38 PM


* I also tested this PR with a RayCluster, and FQ_RAY_IP is as expected.
  Screen Shot 2023-03-20 at 3 29 12 PM

**Comment by kevin85421** [comment]
The CI failure has no relationship with this PR. Merged.

**Pull request merged**
